README.md
heartbleed-masstest
===================

The purpose of this repo is to monitor the vulnerability state of top websites to Heartbleed, so that users may know which websites may or may not have been safe, which passwords should be changed and if their data may have been compromised.

This repo contains a script to automatically test sites for vulnerability to the [Heartbleed Bug (CVE-2014-0160)](http://heartbleed.com/).

*Note: the proof-of-concept has been modified to test for Heartbleed without exploiting the server, so the heartbeat request does not cause the server to leak any data from memory or expose any data in an unauthorized manner. See the [Mozilla blog post](http://blog.mozilla.org/security/2014/04/12/testing-for-heartbleed-vulnerability-without-exploiting-the-server) for an explanation.*

This repo also contains test results for the Alexa top 10,000 sites as of around April 8, 16:00 UTC. Websites listed
as vulnerable may no longer be vulnerable. This list serves as a snapshot of
vulnerable sites at the time of the scan.

To check if a site is still vulnerable, you may use the tool at http://filippo.io/Heartbleed/.

Please note that subdomains aren't tested, so sites that don't have SSL on their main domain will appear as "no SSL".

Overview
--------

[Vulnerable sites on April 8, 16:00 UTC](#630-of-the-top-10000-sites-appeared-vulnerable-on-april-8-1600-utc) | [Changes on April 12, 21:00 UTC](#changes-on-april-12-2100-utc) | [All sites on April 8, 16:00 UTC](#test-results-of-top-1000-websites-on-april-8-1600-utc)

### 630 of the top 10000 sites appeared vulnerable on April 8, 16:00 UTC:

    yahoo.com
    imgur.com
    stackoverflow.com
    kickass.to
    flickr.com
    sogou.com
    adf.ly
    outbrain.com
    archive.org
    addthis.com
    stackexchange.com
    popads.net
    php.net
    avito.ru
    kaskus.co.id
    web.de
    chaturbate.com
    zeobit.com
    beeg.com
    seznam.cz
    okcupid.com
    pch.com
    xda-developers.com
    steamcommunity.com
    scoop.it
    hidemyass.com
    123rf.com
    m-w.com
    dreamstime.com
    majesticseo.com
    amung.us
    duckduckgo.com
    leo.org
    eventbrite.com
    wetransfer.com
    sh.st
    entrepreneur.com
    zoho.com
    yts.re
    usmagazine.com
    fool.com
    letitbit.net
    digitalpoint.com
    picmonkey.com
    petflow.com
    squidoo.com
    avazutracking.net
    elegantthemes.com
    500px.com
    wpmudev.org
    junbi-tracker.com
    gogetlinks.net
    freelotto.com
    cheezburger.com
    weather.gov
    zergnet.com
    city-data.com
    mail.com
    bab.la
    internettrafficformula.com
    startimes.com
    homeshop18.com
    rollingstone.com
    smi2.ru
    gogvo.com
    indiegogo.com
    livetv.sx
    infowars.com
    zap2it.com
    breitbart.com
    fatwallet.com
    internetdownloadmanager.com
    yjc.ir
    chess.com
    doodle.com
    sendspace.com
    similarweb.com
    inc.com
    bomnegocio.com
    studiopress.com
    privatbank.ua
    juicyads.com
    theatlantic.com
    chefkoch.de
    mirtesen.ru
    clickey.com
    letras.mus.br
    tusfiles.net
    reverbnation.com
    digitaltrends.com
    searchenginejournal.com
    torcache.net
    yify-torrents.com
    championat.com
    arstechnica.com
    sme.sk
    zomato.com
    economist.com
    slimspots.com
    jasmin.com
    lonelyplanet.com
    unity3d.com
    androidcentral.com
    wanggou.com
    tubegalore.com
    gi-akademie.com
    rarbg.com
    prezentacya.ru
    watchtower.com
    xxxbunker.com
    kaspersky.com
    adxcore.com
    depositfiles.com
    xtool.ru
    rantlifestyle.com
    xml-sitemaps.com
    banglanews24.com
    worldoftanks.ru
    webpagetest.org
    rapidshare.com
    funnie.st
    vipzona.info
    bizrate.com
    creativecommons.org
    izlesene.com
    therichest.com
    dailycaller.com
    newsnow.co.uk
    webutations.org
    srclick.ru
    tokopedia.com
    minus.com
    cyberciti.biz
    tune.pk
    filmaffinity.com
    skrill.com
    paipai.com
    oschina.net
    fitbit.com
    myanimelist.net
    bidorbuy.co.za
    rantsports.com
    wikispaces.com
    worldoftanks.eu
    cdn4711.net
    hotukdeals.com
    zovi.com
    dfiles.ru
    zone-telechargement.com
    drugs.com
    mediatemple.net
    xmarks.com
    rozetka.com.ua
    appannie.com
    thestreet.com
    arvixe.com
    transfermarkt.de
    worldoftanks.com
    affiliatewindow.com
    bigpoint.com
    zdf.de
    profitcentr.com
    mydealz.de
    ssisurveys.com
    toshiba.com
    indowebster.com
    dsdomination.com
    utorrent.com
    nosub.tv
    filelist.ro
    gazzetta.gr
    medium.com
    imperiaonline.org
    vic.gov.au
    creditkarma.com
    online-convert.com
    myegy.to
    searchfunmoods.com
    fark.com
    scamadviser.com
    bizdec.ru
    bannersbroker.com
    joomlart.com
    webinarjam.net
    pornyaz.com
    unetenet.com
    joyreactor.cc
    globallshare.com
    tuniu.com
    csfd.cz
    uservoice.com
    trafficfactory.biz
    barclaycardus.com
    mci.ir
    noulinx.com
    ingbank.pl
    gonzoxxxmovies.com
    cabelas.com
    khamsat.com
    2gis.ru
    fab.com
    digitalmarketer.com
    billionuploads.com
    efukt.com
    shaparak.ir
    radiojavan.com
    christianpost.com
    pons.com
    ymlp.com
    btcclicks.com
    mythemeshop.com
    seo-fast.ru
    mojo-themes.com
    draugiem.lv
    final.ir
    madmimi.com
    sinaapp.com
    ptcsolution.com
    bytes.com
    attracta.com
    dek-d.com
    dfiles.eu
    graphicstock.com
    europe1.fr
    fineartamerica.com
    text.ru
    dastelefonbuch.de
    enter.ru
    awempire.com
    docnhat.net
    megaindex.ru
    ucla.edu
    wisegeek.com
    flightaware.com
    peerfly.com
    wistia.com
    extremetracking.com
    ghatreh.com
    depositfiles.org
    palcomp3.com
    uiuc.edu
    fancy.com
    tagesanzeiger.ch
    dynadot.com
    soup.io
    ip138.com
    jumia.com.ng
    worthofweb.com
    adage.com
    talktalk.co.uk
    playcast.ru
    congratulations-you-won.com
    ge.tt
    informador.com.mx
    hammihan.com
    onedio.com
    olx.co.th
    gi-backoffice.com
    gorillavid.in
    talkingpointsmemo.com
    518.com.tw
    fast-torrent.ru
    animoto.com
    webinarjam.com
    arioo.com
    gogoanime.com
    mojomarketplace.com
    nanapi.jp
    bahseazad.ir
    serpbook.com
    libsyn.com
    playxn.com
    classifiedads.com
    myus.com
    q.gs
    lijit.com
    joxi.ru
    vbox7.com
    ocj.com.cn
    giga.de
    tomoson.com
    alkislarlayasiyorum.com
    overclock.net
    domainnamesales.com
    appthemes.com
    katestube.com
    lupoporno.com
    windguru.cz
    designmodo.com
    berlin.de
    lifo.gr
    diepresse.com
    postjoint.com
    bittorrent.com
    worldoftanks.asia
    okitspace.com
    sparkasse.at
    nascar.com
    slate.fr
    njuskalo.hr
    astro.com
    lavozdegalicia.es
    androidpolice.com
    longtailvideo.com
    webbirga.net
    idwebgame.com
    gfan.com
    webempresa.com
    bibsonomy.org
    wenyard.com
    readwrite.com
    vmoptions.com
    wowkeren.com
    ay.gy
    ibtimes.co.in
    ss.lv
    coinmill.com
    memecenter.com
    dhnet.be
    ifttt.com
    songmeanings.com
    kicktipp.de
    graphixshare.com
    jango.com
    zoomit.ir
    userscripts.org
    uyan.cc
    down1oads.com
    full-stream.net
    1001freefonts.com
    hotnews.ro
    mightydeals.com
    zimbra.free.fr
    apktops.ir
    selfhtml.org
    shopzilla.com
    dha.com.tr
    copy.com
    coderanch.com
    darty.com
    24tv.ua
    smashwords.com
    castorama.fr
    solidtrustpay.com
    yatedo.com
    neurs.com
    mdir.ir
    thewire.com
    futbol24.com
    keek.com
    elitetorrent.net
    ecosia.org
    j.gs
    jonloomer.com
    deseretnews.com
    spinding.com
    theme123.net
    biography.com
    check24.de
    cplusplus.com
    safecart.com
    healthkart.com
    baskino.com
    cifraclub.com.br
    exactseek.com
    expatriates.com
    saharareporters.com
    feebbo.com
    adshostnet.com
    webartex.ru
    pond5.com
    wpml.org
    dl-protect.com
    seocentro.com
    flexmls.com
    a2hosting.com
    gov.cl
    savenkeep.com
    internetlifestylenetwork.com
    uploadbaz.com
    optimizepress.com
    mttbsystem.com
    pulptastic.com
    theme.co
    swansonvitamins.com
    afterbuy.de
    zorpia.com
    1c-bitrix.ru
    holidaylettings.co.uk
    premiere.fr
    socialadr.com
    sixrevisions.com
    hypovereinsbank.de
    themefuse.com
    roodo.com
    cbr.ru
    marunadanmalayali.com
    jquery4u.com
    powtoon.com
    ubersuggest.org
    markethealth.com
    freelogoservices.com
    readthedocs.org
    mandrillapp.com
    pixeden.com
    nukistream.com
    tractionize.com
    avito.ma
    101domain.com
    sinemalar.com
    weathernews.jp
    nasgo.net
    luxup.ru
    wasanga.com
    realself.com
    matadornetwork.com
    ee.co.uk
    lovoo.net
    paperblog.com
    technobuffalo.com
    zoom.com.br
    bimlatino.com
    dramafever.com
    xojane.com
    azhibo.com
    edgecastcdn.net
    duke.edu
    mobilelikez.com
    cooperativa.cl
    asiatech.ir
    phpnuke.org
    avsforum.com
    99wed.com
    wine-searcher.com
    cyberchimps.com
    umd.edu
    t24.com.tr
    imgsrc.ru
    farnell.com
    phpbb.com
    fermasosedi.ru
    smartinsights.com
    pik.ba
    qq163.com
    pciconcursos.com.br
    hsoub.com
    sixfigurefunnelformula.com
    kingworldnews.com
    heritage.org
    quidco.com
    mediapost.com
    watchcric.com
    vodafone.co.uk
    viki.com
    bplans.com
    all-union.com
    clickprime8.com
    competitor.com
    beliefnet.com
    pingler.com
    worthytoshare.com
    millenium.org
    encuentra24.com
    skimlinks.com
    mubasher.info
    marksdailyapple.com
    buenosaires.gob.ar
    gelbeseiten.de
    lalibre.be
    mobikwik.com
    affilorama.com
    appcelerator.com
    invisionpower.com
    tz4.com
    favstar.fm
    zenhabits.net
    customink.com
    templateism.com
    getsatisfaction.com
    rosnet.ru
    digitalriver.com
    joomlashine.com
    fide.com
    lyricsmode.com
    trovit.it
    socialfabric.us
    rcgroups.com
    directadmin.com
    hrsmart.com
    eprize.com
    qoinpro.com
    path.com
    mobeoffice.com
    taz.de
    authorityroi.com
    jqueryscript.net
    wakeupnow.com
    abakus-internet-marketing.de
    cinepolis.com
    dlisted.com
    joomlaportal.de
    ncsu.edu
    imasters.com.br
    adworkmedia.com
    entropay.com
    unionpaysecure.com
    maxpark.com
    popupads.ir
    getrichradio.com
    proranktracker.com
    androidpit.com
    polki.pl
    ip2location.com
    radiko.jp
    wisegeek.org
    trovit.com
    jobisjob.co.in
    quondos.com
    softportal.com
    nationaljournal.com
    ebesucher.de
    zive.cz
    neurs.net
    telewebion.com
    performancehorizon.com
    springpad.com
    theync.com
    mg.gov.br
    globus-inter.com
    osclass.org
    cpasuperaffiliate.com
    naukrigulf.com
    trovit.com.mx
    eventbrite.co.uk
    paris.fr
    kenrockwell.com
    zagat.com
    444.hu
    barchart.com
    edlen24.com
    peixeurbano.com.br
    flamingtext.com
    speedyshare.com
    seratnews.ir
    bizpowa.com
    grasscity.com
    downloadab.com
    sportdog.gr
    programmableweb.com
    fontpalace.com
    head-fi.org
    bitterstrawberry.com
    responsinator.com
    cuisineaz.com
    work.ua
    utsandiego.com
    foozine.com
    androidpit.de
    invisionzone.com
    top.de
    rememberthemilk.com
    infogr.am
    blizko.ru
    elle.fr
    aremo.com.br
    static.squarespace.com
    fotki.com
    datropy.com
    adscendmedia.com
    spartoo.com
    wordcounter.net
    example.com
    turkcealtyazi.org
    worthytoshare.net
    ideeli.com
    trthaber.com
    romedic.ro
    vudu.com
    aplus.com
    lavozdelmuro.com
    citruspay.com
    optimizehub.com
    hobo-web.co.uk
    dailybasis.com
    sports.fr
    dressupgamesite.com
    smartbrief.com
    scoopwhoop.com
    dev7studios.com
    longurl.it
    privatbank.ru
    wannonce.com
    topnews.ru
    55188.com
    cinepolis.com.mx
    gorilla-evolution.com
    keyandway.com
    perfectworld.eu
    shopware.de
    kyivpost.com
    bavotasan.com
    gnetwork.biz
    farsisubtitle.com
    kat.ph
    vocabulary.com
    daveramsey.com
    mumsnet.com
    alfajertv.com
    gordonua.com
    friendorfollow.com
    peeplo.com
    free-ebooks.net
    adslzone.net
    adultbay.org
    softcoin.com
    trovit.es
    evsuite.com
    waseet.net
    sudaneseonline.com
    mobypicture.com
    pinkbike.com
    fizzle.co
    telly.com
    blip.tv
    basketball-reference.com
    resellerratings.com
    quirktools.com
    afrihost.com
    18qt.com
    winscp.net

### Changes on April 12, 21:00 UTC ###
| Website              | 04/08/2014           |  04/12/2014           | 
|----------------------|----------------------|-----------------------|
| 1001freefonts.com    | vulnerable.          |  not vulnerable.      | 
| 101domain.com        | vulnerable.          |  not vulnerable.      | 
| 11st.co.kr           | not vulnerable.      |  no SSL.              | 
| 123rf.com            | vulnerable.          |  not vulnerable.      | 
| 18qt.com             | vulnerable.          |  not vulnerable.      | 
| 1c-bitrix.ru         | vulnerable.          |  not vulnerable.      | 
| 1fichier.com         | no SSL.              |  not vulnerable.      | 
| 24tv.ua              | vulnerable.          |  not vulnerable.      | 
| 2ch-c.net            | not vulnerable.      |  no SSL.              | 
| 2gis.ru              | vulnerable.          |  not vulnerable.      | 
| 444.hu               | vulnerable.          |  not vulnerable.      | 
| 500px.com            | vulnerable.          |  not vulnerable.      | 
| 518.com.tw           | vulnerable.          |  not vulnerable.      | 
| 51bi.com             | not vulnerable.      |  no SSL.              | 
| 56.com               | not vulnerable.      |  no SSL.              | 
| 99wed.com            | vulnerable.          |  not vulnerable.      | 
| a2hosting.com        | vulnerable.          |  not vulnerable.      | 
| abakus-internet-mark | vulnerable.          |  not vulnerable.      | 
| academic.ru          | no SSL.              |  not vulnerable.      | 
| adage.com            | vulnerable.          |  not vulnerable.      | 
| addthis.com          | vulnerable.          |  not vulnerable.      | 
| adevarul.ro          | not vulnerable.      |  vulnerable.          | 
| adf.ly               | vulnerable.          |  not vulnerable.      | 
| adscendmedia.com     | vulnerable.          |  not vulnerable.      | 
| adshostnet.com       | vulnerable.          |  not vulnerable.      | 
| adslzone.net         | vulnerable.          |  not vulnerable.      | 
| adworkmedia.com      | vulnerable.          |  not vulnerable.      | 
| adxcore.com          | vulnerable.          |  not vulnerable.      | 
| affiliatewindow.com  | vulnerable.          |  not vulnerable.      | 
| affilorama.com       | vulnerable.          |  not vulnerable.      | 
| afrihost.com         | vulnerable.          |  not vulnerable.      | 
| afterbuy.de          | vulnerable.          |  not vulnerable.      | 
| alfajertv.com        | vulnerable.          |  not vulnerable.      | 
| alimama.com          | not vulnerable.      |  no SSL.              | 
| alkislarlayasiyorum. | vulnerable.          |  not vulnerable.      | 
| alluc.to             | not vulnerable.      |  no SSL.              | 
| amazon.cn            | not vulnerable.      |  no SSL.              | 
| amung.us             | vulnerable.          |  not vulnerable.      | 
| ana.co.jp            | not vulnerable.      |  no SSL.              | 
| androidcentral.com   | vulnerable.          |  not vulnerable.      | 
| androidpit.com       | vulnerable.          |  not vulnerable.      | 
| androidpit.de        | vulnerable.          |  not vulnerable.      | 
| androidpolice.com    | vulnerable.          |  not vulnerable.      | 
| animenewsnetwork.com | not vulnerable.      |  no SSL.              | 
| animoto.com          | vulnerable.          |  not vulnerable.      | 
| anoox.com            | not vulnerable.      |  vulnerable.          | 
| antarvasna.com       | no SSL.              |  not vulnerable.      | 
| apktops.ir           | vulnerable.          |  not vulnerable.      | 
| aplus.com            | vulnerable.          |  not vulnerable.      | 
| appannie.com         | vulnerable.          |  not vulnerable.      | 
| appcelerator.com     | vulnerable.          |  not vulnerable.      | 
| appthemes.com        | vulnerable.          |  not vulnerable.      | 
| archive.org          | vulnerable.          |  not vulnerable.      | 
| arstechnica.com      | vulnerable.          |  not vulnerable.      | 
| arvixe.com           | vulnerable.          |  not vulnerable.      | 
| assembla.com         | not vulnerable.      |  no SSL.              | 
| astro.com            | vulnerable.          |  not vulnerable.      | 
| attracta.com         | vulnerable.          |  not vulnerable.      | 
| authorityroi.com     | vulnerable.          |  not vulnerable.      | 

changes.md
| Website              | 04/08/2014           |  04/11/2014           | 
|----------------------|----------------------|-----------------------|
| 1001freefonts.com    | vulnerable.          |  not vulnerable.      | 
| 101domain.com        | vulnerable.          |  not vulnerable.      | 
| 123rf.com            | vulnerable.          |  not vulnerable.      | 
| 163.com              | no SSL.              |  not vulnerable.      | 
| 18qt.com             | vulnerable.          |  not vulnerable.      | 
| 1c-bitrix.ru         | vulnerable.          |  not vulnerable.      | 
| 1fichier.com         | no SSL.              |  not vulnerable.      | 
| 24tv.ua              | vulnerable.          |  not vulnerable.      | 
| 2gis.ru              | vulnerable.          |  not vulnerable.      | 
| 444.hu               | vulnerable.          |  not vulnerable.      | 
| 500px.com            | vulnerable.          |  not vulnerable.      | 
| 518.com.tw           | vulnerable.          |  not vulnerable.      | 
| 51bi.com             | not vulnerable.      |  no SSL.              | 
| 99wed.com            | vulnerable.          |  not vulnerable.      | 
| a2hosting.com        | vulnerable.          |  not vulnerable.      | 
| abakus-internet-mark | vulnerable.          |  not vulnerable.      | 
| academic.ru          | no SSL.              |  not vulnerable.      | 
| adage.com            | vulnerable.          |  not vulnerable.      | 
| addthis.com          | vulnerable.          |  not vulnerable.      | 
| adf.ly               | vulnerable.          |  not vulnerable.      | 
| adscendmedia.com     | vulnerable.          |  not vulnerable.      | 
| adshostnet.com       | vulnerable.          |  not vulnerable.      | 
| adslzone.net         | vulnerable.          |  not vulnerable.      | 
| adworkmedia.com      | vulnerable.          |  not vulnerable.      | 
| adxcore.com          | vulnerable.          |  not vulnerable.      | 
| affiliatewindow.com  | vulnerable.          |  not vulnerable.      | 
| affilorama.com       | vulnerable.          |  not vulnerable.      | 
| afrihost.com         | vulnerable.          |  not vulnerable.      | 
| afterbuy.de          | vulnerable.          |  not vulnerable.      | 
| alkislarlayasiyorum. | vulnerable.          |  not vulnerable.      | 
| alriyadh.com         | not vulnerable.      |  no SSL.              | 
| amung.us             | vulnerable.          |  not vulnerable.      | 
| androidcentral.com   | vulnerable.          |  not vulnerable.      | 
| androidpit.com       | vulnerable.          |  not vulnerable.      | 
| androidpit.de        | vulnerable.          |  not vulnerable.      | 
| androidpolice.com    | vulnerable.          |  not vulnerable.      | 
| animoto.com          | vulnerable.          |  not vulnerable.      | 
| anoox.com            | not vulnerable.      |  vulnerable.          | 
| antarvasna.com       | no SSL.              |  not vulnerable.      | 
| apktops.ir           | vulnerable.          |  not vulnerable.      | 
| aplus.com            | vulnerable.          |  not vulnerable.      | 
| appannie.com         | vulnerable.          |  not vulnerable.      | 
| appcelerator.com     | vulnerable.          |  not vulnerable.      | 
| appthemes.com        | vulnerable.          |  not vulnerable.      | 
| archive.org          | vulnerable.          |  not vulnerable.      | 
| arstechnica.com      | vulnerable.          |  not vulnerable.      | 
| arvixe.com           | vulnerable.          |  not vulnerable.      | 
| astro.com            | vulnerable.          |  not vulnerable.      | 
| attracta.com         | vulnerable.          |  not vulnerable.      | 
| authorityroi.com     | vulnerable.          |  not vulnerable.      | 
| avazutracking.net    | vulnerable.          |  not vulnerable.      | 
| avito.ma             | vulnerable.          |  not vulnerable.      | 
| avito.ru             | vulnerable.          |  not vulnerable.      | 
| avsforum.com         | vulnerable.          |  no SSL.              | 
| awempire.com         | vulnerable.          |  not vulnerable.      | 
| ay.gy                | vulnerable.          |  not vulnerable.      | 
| azhibo.com           | vulnerable.          |  no SSL.              | 
| bab.la               | vulnerable.          |  not vulnerable.      | 
| backpage.com         | not vulnerable.      |  no SSL.              | 
| bahseazad.ir         | vulnerable.          |  not vulnerable.      | 
| bannersbroker.com    | vulnerable.          |  not vulnerable.      | 
| barchart.com         | vulnerable.          |  not vulnerable.      | 
| barclaycardus.com    | vulnerable.          |  not vulnerable.      | 
| bartarinha.ir        | no SSL.              |  not vulnerable.      | 
| basketball-reference | vulnerable.          |  not vulnerable.      | 
| baskino.com          | vulnerable.          |  not vulnerable.      | 
| bavotasan.com        | vulnerable.          |  not vulnerable.      | 
| beeg.com             | vulnerable.          |  not vulnerable.      | 
| beitaichufang.com    | no SSL.              |  not vulnerable.      | 
| berlin.de            | vulnerable.          |  not vulnerable.      | 
| bibsonomy.org        | vulnerable.          |  not vulnerable.      | 
| bidorbuy.co.za       | vulnerable.          |  not vulnerable.      | 
| bigpoint.com         | vulnerable.          |  not vulnerable.      | 
| billboard.com        | not vulnerable.      |  no SSL.              | 
| billionuploads.com   | vulnerable.          |  not vulnerable.      | 
| bimlatino.com        | vulnerable.          |  not vulnerable.      | 
| biography.com        | vulnerable.          |  not vulnerable.      | 
| bitterstrawberry.com | vulnerable.          |  not vulnerable.      | 
| bittorrent.com       | vulnerable.          |  not vulnerable.      | 
| bizdec.ru            | vulnerable.          |  not vulnerable.      | 

ssltest.py
#!/usr/bin/env python

# Based on the original code by Jared Stafford.

# NOTE: this code has been modified to test for OpenSSL versions vulnerable to 
# Heartbleed without exploiting the server, therefore the heartbeat request
# does _not_ cause the server to leak any data from memory or expose any data
# in an unauthorized manner.
# Based on: https://github.com/dchan/metasploit-framework/blob/master/modules/auxiliary/scanner/ssl/openssl_heartbleed.rb
# See: https://blog.mozilla.org/security/2014/04/12/testing-for-heartbleed-vulnerability-without-exploiting-the-server/

# Usage example: python ssltest.py example.com

import sys
import struct
import socket
import time
import select
import re
import threading
import netaddr
import json
import os
import datetime
import signal
from optparse import OptionParser
from collections import defaultdict
from multiprocessing.dummy import Pool

host_status = {}
hosts_to_skip = []
counter = defaultdict(int)
lock = threading.Lock()

options = OptionParser(usage='%prog <network> [network2] [network3] ...', description='Test for SSL heartbleed vulnerability (CVE-2014-0160) on multiple domains')
options.add_option('--input', '-i', dest="input_file", default=[], action="append", help="Optional input file of networks or ip addresses, one address per line")
options.add_option('--logfile', '-o', dest="log_file", default="results.txt", help="Optional logfile destination")
options.add_option('--resume', dest="resume", action="store_true", default=False, help="Do not rescan hosts that are already in the logfile")
options.add_option('--timeout', '-t', dest="timeout", default=5, help="How long to wait for remote host to respond before timing out")
options.add_option('--threads', dest="threads", default=100, help="If specific, run X concurrent threads")
options.add_option('--json', dest="json_file", default=None, help="Save data as json into this file")
options.add_option('--only-vulnerable', dest="only_vulnerable", action="store_true", default=False, help="Only scan hosts that have been scanned before and were vulnerable")
options.add_option('--only-unscanned', dest="only_unscanned", action="store_true", default=False, help="Only scan hosts that appear in the json file but have not been scanned")
options.add_option('--summary', dest="summary", action="store_true", default=False, help="Useful with --json. Don't scan, just print old results")
options.add_option('--verbose', dest="verbose", action="store_true", default=False, help="Print verbose information to screen")
options.add_option('--max', dest="max", default=None, help="Exit program after scanning X hosts. Useful with --only-unscanned")
opts, args = options.parse_args()

threadpool = Pool(processes=int(opts.threads))


def h2bin(x):
    return x.replace(' ', '').replace('\n', '').decode('hex')

hello = h2bin('''
16 03 03 00  dc 01 00 00 d8 03 03 53
43 5b 90 9d 9b 72 0b bc  0c bc 2b 92 a8 48 97 cf
bd 39 04 cc 16 0a 85 03  90 9f 77 04 33 d4 de 00
00 66 c0 14 c0 0a c0 22  c0 21 00 39 00 38 00 88
00 87 c0 0f c0 05 00 35  00 84 c0 12 c0 08 c0 1c
c0 1b 00 16 00 13 c0 0d  c0 03 00 0a c0 13 c0 09
c0 1f c0 1e 00 33 00 32  00 9a 00 99 00 45 00 44
c0 0e c0 04 00 2f 00 96  00 41 c0 11 c0 07 c0 0c
c0 02 00 05 00 04 00 15  00 12 00 09 00 14 00 11
00 08 00 06 00 03 00 ff  01 00 00 49 00 0b 00 04
03 00 01 02 00 0a 00 34  00 32 00 0e 00 0d 00 19
00 0b 00 0c 00 18 00 09  00 0a 00 16 00 17 00 08
00 06 00 07 00 14 00 15  00 04 00 05 00 12 00 13
00 01 00 02 00 03 00 0f  00 10 00 11 00 23 00 00
00 0f 00 01 01                                  
''')

def recvall(s, length, timeout=5):
    endtime = time.time() + timeout
    rdata = ''
    remain = length
    while remain > 0:
        rtime = endtime - time.time()
        if rtime < 0:
            return None
        r, w, e = select.select([s], [], [], 5)
        if s in r:
            try:
                data = s.recv(remain)
            except Exception, e:
                return None
            # EOF?
            if not data:
                return None
            rdata += data
            remain -= len(data)
    return rdata


def recvmsg(s):
    hdr = recvall(s, 5)
    if hdr is None:
        return None, None, None
    typ, ver, ln = struct.unpack('>BHH', hdr)
    pay = recvall(s, ln, 10)
    if pay is None:
        return None, None, None
    return typ, ver, pay


def hit_hb(s):
    while True:
        typ, ver, pay = recvmsg(s)
        if typ is None:
            return False

        if typ == 24:
            return True

        if typ == 21:
            return False


def is_vulnerable(host, timeout):
    """ Check if remote host is vulnerable to heartbleed

     Returns:
        None  -- If remote host has no ssl
        False -- Remote host has ssl but likely not vulnerable
        True  -- Remote host might be vulnerable
    """
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.settimeout(int(timeout))
    try:
        s.connect((host, 443))
    except Exception, e:
        return None
    s.send(hello)

    while True:
        typ, ver, pay = recvmsg(s)
        if typ is None:
            return None

        # Look for server hello done message.
        if typ == 22 and ord(pay[0]) == 0x0E:
            break

    # construct heartbeat request packet
    ver_chr = chr(ver&0xff)
    hb  = h2bin("18 03") + ver_chr + h2bin("40 00 01 3f fd") + "\x01"*16381
    hb += h2bin("18 03") + ver_chr + h2bin("00 03 01 00 00")

    s.send(hb)
    return hit_hb(s)


def store_results(host_name, current_status):
    current_time = time.time()
    with lock:
        counter[current_status] += 1
        counter["Total"] += 1
        if host_name not in host_status:
            host_status[host_name] = {}
        host = host_status[host_name]
        # Make a note when this host was last scanned
        host['last_scan'] = current_time

        # Make a note if this host has never been scanned before
        if 'first_scan' not in host:
            host['first_scan'] = current_time
        elif host.get('status', 'never been scanned') != current_status:
            # If it has a different check result from before
            host['changelog'] = host.get('changelog', [])
            changelog_entry = [current_time, current_status]
            host['changelog'].append(changelog_entry)
        host['status'] = current_status
        with open(opts.log_file, 'a') as f:
            message = "{current_time} {host} {current_status}".format(**locals())
            f.write(message + "\n")
            return message


def scan_host(host):
    """ Scans a single host, logs into

    Returns:
        list(timestamp, ipaddress, vulnerabilitystatus)
    """
    if opts.max and int(opts.max) >= counter["Total"]:
        return
    host = str(host)
    if host in hosts_to_skip:
        return
    result = is_vulnerable(host, opts.timeout)
    message = store_results(host, result)
    if opts.verbose:
        print message
    return message


def scan_hostlist(hostlist, threads=5):
    """ Iterates through hostlist and scans them

    Arguments:
        hostlist    -- Iterable with ip addresses
        threads     -- If specified, run in multithreading mode
    """
    task = threadpool.map_async(scan_host, hostlist)
    while True:
        print counter['Total'], "hosts done"
        task.wait(1)
        if task.ready() or hasattr(threadpool, 'done'):
            return
    threadpool.close()
    threadpool.join()


def clean_hostlist(args):
    """ Returns list of iterables
    Examples:
    >>> hostlist = ["127.0.0.1", "127.0.0.2"]
    >>> clean_hostlist(hostlist)
    """
    hosts = []
    networks = []
    for i in args:
        # If it contains any alphanumerics, it might be a domain name
        if any(c.isalpha() for c in i):
            # Special hack, because alexa top x list is kind of weird
            i = i.split('/')[0]
            hosts.append(i)
        # If arg contains a / we assume its a network name
        elif '/' in i:
            networks.append(netaddr.IPNetwork(i))
        else:
            hosts.append(i)
    result = []
    for i in networks:
        result.append(i)
    if hosts:
        result.append(hosts)
    return result


def import_json(filename):
    """ Reads heartbleed data in json format from this file """
    with open(filename) as f:
        json_data = f.read()
    data = json.loads(json_data)
    for k, v in data.items():
        host_status[k] = v


def export_json(filename):
    """ Save scan results into filename as json data
    """
    json_data = json.dumps(host_status, indent=4)
    with open(filename, 'w') as f:
        f.write(json_data)


def print_summary():
    """ Print summary of previously stored json data to screen """
    if not opts.json_file:
        pass
        #options.error("You need to provide --json with --summary")
    else:
        import_json(opts.json_file)
    counter = defaultdict(int)
    for host, data in host_status.items():
        friendly_status = "unknown"
        status = data.get('status', "Not scanned")
        if status is None:
            friendly_status = "SSL Connection Failed"
        elif status is True:
            friendly_status = "Vulnerable"
        elif status is False:
            friendly_status = "Not Vulnerable"
        else:
            friendly_status = str(status)
        last_scan = int(float(data.get('last_scan',0)))
        last_scan = datetime.datetime.fromtimestamp(last_scan).strftime('%Y-%m-%d %H:%M:%S')
        counter[friendly_status] += 1
        counter['Total'] += 1
        if opts.only_vulnerable and not status:
            continue
        elif opts.only_unscanned and 'status' in data:
            continue
        print "%s %-20s %5s" % (last_scan, host, friendly_status)
    print "------------ summary -----------"
    for k,v in counter.items():
        print "%-7s %s" % (v, k)
    return

def signal_handler(signal, frame):
    print "Ctrl+C pressed.. aborting..."
    threadpool.terminate()
    threadpool.done = True

def main():
    if opts.summary:
        print_summary()
        return

    if not args and not opts.input_file and not opts.json_file:
        options.print_help()
        return


    # If any input files were provided, parse through them and add all addresses to "args"
    for input_file in opts.input_file:
        with open(input_file) as f:
            for line in f:
                words = line.split()
                if not words:
                    continue
                # If input file is in masscan's portscan format
                if line.startswith("Discovered open port"):
                    args.append(words.pop())
                elif len(words) == 1:
                    args.append(words[0])
                else:
                    print "Skipping invalid input line: " % line
                    continue
    if opts.json_file:
        try:
            import_json(opts.json_file)
        except IOError:
            print opts.json_file, "not found. Not importing any data"


        for host_name, data in host_status.items():
            if opts.only_unscanned and 'status' in data:
                continue
            if data.get('status', None) is True or not opts.only_vulnerable:
                args.append(host_name)

    # For every network in args, convert it to a netaddr network, so we can iterate through each host
    remote_networks = clean_hostlist(args)
    for network in remote_networks:
        scan_hostlist(network, threads=opts.threads)

    if opts.json_file:
        export_json(opts.json_file)

    print_summary()

if __name__ == '__main__':
    signal.signal(signal.SIGINT, signal_handler)
    main()
